Add breadcrumb contextmenu

pascalwengerter opened the pull request:
## Related Issue
- Works towards https://github.com/owncloud/web/issues/6030

## How Has This Been Tested?
- Manually

## Types of changes
- [X] New feature (non-breaking change which adds functionality)
- [ ] Tests

## Checklist:
- [X] Code changes
- [x] Unit tests added/updated
- [X] Documentation updated

pascalwengerter commented:
@JanAckermann @JammingBen feel free to update/review/merge this one if it suits your needs in `web`!

JammingBen commented:
@pascalwengerter It's working just fine 👍  Some tests here would be nice though, what do you think?

pascalwengerter commented:
> @pascalwengerter It's working just fine +1 Some tests here would be nice though, what do you think?

Yes please. Checking that the slot doesn't get displayed if the value is set to false or if it is set to true but only one breadcrumb item is handed over (or sth like that) 👌🏽

JammingBen commented:
@pascalwengerter I updated the PR with unit tests and added `showContextmenu` as computed property. Please [have a look](https://github.com/owncloud/owncloud-design-system/pull/1786/commits/aab7c451372daae9e7857de329ed3d85d93ae834), since you probably can't review your own PR.

pascalwengerter commented:
> @pascalwengerter I updated the PR with unit tests and added `showContextmenu` as computed property. Please [have a look](https://github.com/owncloud/owncloud-design-system/pull/1786/commits/aab7c451372daae9e7857de329ed3d85d93ae834), since you probably can't review your own PR.

Can't approve but LGTM ;)

kulmann commented on src/components/molecules/OcBreadcrumb/OcBreadcrumb.vue:
```diff
@@ -110,6 +131,12 @@ export default {
+    contextMenuLabel() {
+      return this.$gettext("Show actions for current folder")
+    },
+    showContextmenu() {
```
```suggestion
    showContextMenu() {
```
for consistency with the computed prop above.

kulmann commented on src/components/molecules/OcBreadcrumb/OcBreadcrumb.vue:
```diff
@@ -14,6 +14,27 @@
           {{ item.text }}
         </oc-button>
         <span v-else :aria-current="getAriaCurrent(index)" tabindex="-1" v-text="item.text" />
+        <template v-if="showContextmenu && index == items.length - 1 && items.length > 1">
```
```suggestion
        <template v-if="showContextmenu && index === items.length - 1 && items.length > 1">
```
I think the last check (`items.length > 1`) belongs in web, because if we'd introduce actions for the root node we would not be able to do so with this implementation.

kulmann commented on src/components/molecules/OcBreadcrumb/OcBreadcrumb.spec.js:
```diff
@@ -20,15 +20,49 @@ describe("OcBreadcrumb", () => {
+      },
+    })
+
+    expect(wrapper.findAll(".oc-breadcrumb-list-item").length).toBe(4)
```
The static check for `4` elements makes this test quite brittle. Can you check with `items.length`?

kulmann commented on src/components/molecules/OcBreadcrumb/OcBreadcrumb.spec.js:
```diff
@@ -20,15 +20,49 @@ describe("OcBreadcrumb", () => {
+      },
+    })
+
+    expect(wrapper.findAll("#oc-breadcrumb-contextmenu-trigger").length).toBe(1)
```
```suggestion
    expect(wrapper.find("#oc-breadcrumb-contextmenu-trigger").exists()).toBe(true)
```

kulmann commented on src/components/molecules/OcBreadcrumb/OcBreadcrumb.spec.js:
```diff
@@ -20,15 +20,49 @@ describe("OcBreadcrumb", () => {
+    expect(wrapper.findAll("#oc-breadcrumb-contextmenu-trigger").length).toBe(0)
+    expect(wrapper).toMatchSnapshot()
+  })
+  it("displays context menu trigger", () => {
```
Could you group all the context menu tests into a `describe('context menu', () => {` please?

kulmann replied:
```suggestion
  it("displays context menu trigger when a slot is given", () => {
```

kulmann commented on src/components/molecules/OcBreadcrumb/OcBreadcrumb.spec.js:
```diff
@@ -20,15 +20,49 @@ describe("OcBreadcrumb", () => {
+    })
+
+    expect(wrapper.findAll(".oc-breadcrumb-list-item").length).toBe(4)
+    expect(wrapper.findAll("#oc-breadcrumb-contextmenu-trigger").length).toBe(0)
```
Line is not needed because you have a dedicated test for this (`does not display context menu trigger when no slot given` some lines below). Don't test things multiple times. ;-)

kulmann commented on src/components/molecules/OcBreadcrumb/OcBreadcrumb.spec.js:
```diff
@@ -20,15 +20,49 @@ describe("OcBreadcrumb", () => {
 
-      expect(wrapper.findAll(".oc-breadcrumb-list-item").length).toBe(4)
-      expect(wrapper).toMatchSnapshot()
+    expect(wrapper.findAll("#oc-breadcrumb-contextmenu-trigger").length).toBe(0)
```
```suggestion
    expect(wrapper.find("#oc-breadcrumb-contextmenu-trigger").exists()).toBe(false)
```

kulmann commented on src/components/molecules/OcBreadcrumb/OcBreadcrumb.spec.js:
```diff
@@ -20,15 +20,49 @@ describe("OcBreadcrumb", () => {
-      expect(wrapper).toMatchSnapshot()
+    expect(wrapper.findAll("#oc-breadcrumb-contextmenu-trigger").length).toBe(0)
+  })
+  it("does not display context menu trigger with only one item", () => {
```
Test not needed if you think that my comment about the length check is a good idea.

kulmann commented on src/components/molecules/OcBreadcrumb/OcBreadcrumb.spec.js:
```diff
@@ -20,15 +20,36 @@ describe("OcBreadcrumb", () => {
+        items,
+      },
     })
+    console.log(wrapper.html())
```
dev leftover? ;-)

JammingBen replied:
Ahh dammit :)

kulmann reviewed (approved):
Awesome! 😍
